### DISCUSSION TITLE
TestClient follow_redirects

### DISCUSSION THREAD

**Discussion by aminalaee** [encode/starlette#1871] [Potential Issue] [answered]
I think before https://github.com/encode/starlette/pull/1376 the TestClient would not follow redirects,
but it's now set `follow_redirects=True` by default.

I'm not sure if this was on purpose or by mistake. Other than failing existing tests, now it has a different behaviour compared to httpx because httpx has follow_redirects=False by default.

Here's a sample code to show the difference:

```py
from starlette.applications import Starlette
from starlette.routing import Route
from starlette.responses import PlainTextResponse, RedirectResponse
from starlette.testclient import TestClient
from httpx import AsyncClient


def foo(request):
    return PlainTextResponse("Hello World!")


def bar(request):
    return RedirectResponse(request.url_for("foo"), status_code=302)


app = Starlette(
    routes=[
        Route("/foo", foo),
        Route("/bar", bar),
    ]
)


async def test_async():
    client = AsyncClient(app=app, base_url="http://testserver")
    response = await client.get("/bar")
    assert response.status_code == 302


def test_sync():
    client = TestClient(app)
    response = client.get("/bar")
    assert response.status_code == 200
```

I found this today as I saw pipelines failing for my project: https://github.com/aminalaee/sqladmin which use both sync and async tests.

**Comment by aminalaee** [comment]
@Kludex Any ideas?

**Reply by Kludex** [reply]
I'm aware of the different behavior of the HTTP clients, but I think we missed that. We didn't decide that.

I'm unsure if we want to fix this. What do you think?

**Reply by aminalaee** [reply]
I think this is not an urgent fix, but anyway it would be nicer to not break things in users code if we don't really want to.

**Reply by aminalaee** [reply]
I can propose a PR so we can get a review and decide.

**Reply by Kludex** [reply]
I'd appreciate. Thanks Amin 🙏 

We should probably confirm this was actually not thought on the PR that introduced it.

**Reply by aminalaee** [reply]
Yeah sure.

**Reply by Kludex** [reply]
What was the conclusion here? :thinking:

**Reply by aminalaee** [reply]
I think that we are in a better position than before (?)
https://github.com/encode/starlette/pull/1872#issuecomment-1259598223

**Reply by inikolaev** [reply]
What was the conclusion after all? I'm running into this issue and what I need is to disable following redirects on `TestClient` level. I'm fine if this behaviour is enabled by default, but please provide means to configure that. Right now it can only be configured for each individual request.

Something like this would work perfectly:

```python
client = TestClient(follow_redirects=False)
```

**Comment by Kludex** [comment, marked as answer]
This was the conclusion: https://github.com/encode/starlette/pull/1872#issuecomment-1259598223

**Reply by inikolaev** [reply]
Would it be acceptable make that behaviour configurable? Right now I basically had to make a wrapper to make it work for our use-case:

```python
class TestClientWrapper(TestClient):
    """TestClient wrapper with disabled redirects."""

    def _choose_redirect_arg(
        self,
        follow_redirects: Optional[bool],
        allow_redirects: Optional[bool],
    ) -> Union[bool, httpx._client.UseClientDefault]:
        return False

```
